File: Section11_Sensor_Fusion/bumperbot_ws/src/bumperbot_localization/bumperbot_localization/imu_republisher.py
#!/usr/bin/env python3
import rclpy
import time
from rclpy.node import Node
from sensor_msgs.msg import Imu

imu_pub = None

def imuCallback(imu):
    global imu_pub
    imu.header.frame_id = "base_footprint_ekf"
    imu_pub.publish(imu)


def main(args=None):
    global imu_pub
    rclpy.init(args=args)
    node = Node('imu_republisher_node')
    time.sleep(1)
    imu_pub = node.create_publisher(Imu, "imu_ekf", 10)
    imu_sub = node.create_subscription(Imu, "imu/out", imuCallback, 10)
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Section12_Build_the_Robot/bumperbot_ws/src/bumperbot_localization/src/imu_republisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>

using namespace std::chrono_literals;

rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub;

void imuCallback(const sensor_msgs::msg::Imu &imu)
{
  sensor_msgs::msg::Imu new_imu;
  new_imu = imu;
  new_imu.header.frame_id = "base_footprint_ekf";
  imu_pub->publish(new_imu);
}


int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("imu_republisher_node");
  rclcpp::sleep_for(1s);
  imu_pub = node->create_publisher<sensor_msgs::msg::Imu>("imu_ekf", 10);
  auto imu_sub = node->create_subscription<sensor_msgs::msg::Imu>("imu/out", 10, imuCallback);
  
  rclcpp::spin(node);
  rclcpp::shutdown();
}